entry_proxy/proxy_test.go

```diff
@@ -0,0 +1,31 @@
+package main
+
+import (
+	"net"
+	"testing"
+)
+
+type MockSNIParser struct{}
+
+func (m MockSNIParser) ServerNameFromConn(clientConn net.Conn) (string, net.Conn, error) {
+	return "Horse25519", clientConn, nil
+}
+
+type MockOnionResolver struct{}
+
+func (o *MockOnionResolver) ResolveToOnion(hostname string) (string, error) {
+	return "MockOnionHaHa", nil
+}
+
+func TestTLSProxy(t *testing.T) {
+	proxyNet := "tcp"
+	proxyAddr := "127.0.0.1:34627"
+
+	proxy := NewTLSProxy(443, proxyNet, proxyAddr)
```

Comment:
It fails if 34627 port is not free.
See how `proxy_test.go` [is written](https://go.googlesource.com/net/+/master/proxy/proxy_test.go#53):
```go
endSystem, err := net.Listen("tcp", "127.0.0.1:0")
// ...
if c, err := proxy.Dial("tcp", endSystem.Addr().String()); err != nil {
```